DRAFT: CDK: High level architecture (#1563)
0xPolygon/polygon-docs: docs/cdk/architecture/high-level-views.md

## CDK full execution proof (FEP)

The following diagram depicts the CDK FEP stack component layout and details some of the high level communication between components.

![High level view of CDK stack](../../img/cdk/cdk-stack.png)

### Component interactions

- External applications interact with the CDK Erigon RPC node via the transaction pool manager.
- Engineers use a CLI tool to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.
- The CDK Erigon RPC sends transactions to the CDK Erigon sequencer which sequences transactions batches.
- The CDK Erigon sequencer puts the transaction batches into the data streamer.

> - The data streamer streams data to the sequencer sender (validium mode) and the aggregator.

Reviewer: I think that (validium mode) can be removed. In both modes send the same information

- The sequencer sender sequences batches into the L1 smart contract domain and persists data into DAC nodes for validium mode operations.

> - The aggregator sends batches to the prover and receives proofs in return. It then aggregates the proofs into batches before submitting them to the L1 domain.

Reviewer: The aggregator send proofs to AggLayer (maybe `L1 domain` have the same meaning...)

Reviewer 2: ```suggestion
- The aggregator sends batches to the prover and receives proofs in return. It then aggregates the proofs into batches before submitting them to the Agglayer or L1 depending on the chosen settlement layer.
```

How about this?

- Users interact with the bridge service via the bridge UI.
- The AggLayer also utilizes the bridge service and interacts with the L1 smart contracts.

### User data flow

The following diagram is a sequential depiction of the user data flow for the CDK FEP stack in validium mode using a mock prover and having an AggLayer connection.

![High level view of CDK user data flow](../../img/cdk/user-data-flow.png)

#### Sequential interactions

1. User sends a transaction to the CDK Erigon RPC node.
2. The Erigon RPC node proxies the data to the CDK Erigon sequencer node.
3. The sequencer sequences the transaction batches.
4. The sequencer puts the batches into the data streamer component.

> 5. The data streamer component streams the data into the sequence sender for validium mode stacks. It streams data into the aggregator and AggLayer for rollup and validium mode stacks. 

Reviewer: The same as before, doesn't mind if it's or not validium

6. In validium mode, the sequencer sender persists transaction data into the DAC nodes.

> 7. In validium mode, the sequencer sender sequences the batches into the L1 smart contracts.

Reviewer: That is done in both modes: validium and rollup

8. In both modes, the aggregator sends the batches to the prover and the prover returns the proofs.
9. The aggregator batches the proofs.
10. The aggregator submits the final proof to the AggLayer.

> 11. The AggLayer submits the final proof to the L1 smart contract domain.

Reviewer 3: Would it make sense to introduce some sequence diagram, that depicts this process?

## CDK full execution proof (FEP)

> The following diagram depicts the CDK FEP stack component layout and details some of the high level communication between components.

Reviewer 4: ```suggestion
The diagram below depicts a simplified architectural layout of the CDK FEP stack and indicates at a high level how components communicate.
```

![High level view of CDK stack](../../img/cdk/cdk-stack.svg)

### Component interactions

> - Engineers use a CLI to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.

Reviewer 4: ```suggestion
- The CLI tool is the starting point. Developers, or chain administrators in particular, can use the CLI tool to build and configure chains in various modes of operation, such as validiums and rollups.
```

> - External applications send transactions to the CDK Erigon RPC node which forwards the transaction data to the sequencer via the transaction pool manager.

Reviewer 4: ```suggestion
- Once a chain is configured with the CLI, users can submit transactions through the CDK Erigon RPC node. These transactions are relayed to the `tx-pool manager` before the sequencer selects and executes them.
```

> - The sequencer sequences transactions batches and synchronizes data with the RPC node.

Reviewer 4: ```suggestion
- The CDK Erigon sequencer executes the transactions, puts them in blocks and the blocks fill up batches. The CDK Erigon RPC node syncs with the CDK Erigon sequencer.
```

> - The sequencer sender reads batch data from the RPC node.

Reviewer 4: ```suggestion
- The sequencer sender reads batch data from the CDK Erigon RPC node.
```

- The aggregator reads batch data from the sequencer data stream.

> - The sequencer sender persists data into the L1 smart contract domain for rollup mode and into DAC nodes for validium mode operations.

Reviewer 4: ```suggestion
- Depending on whether the chain is a rollup or validium, the sequencer sender sends batch data to either smart contracts on L1 or a DAC.
```

> - The aggregator sends batches to the prover and receives proofs in return. Together with the prover, it aggregates the proofs into batches before submitting the final proofs to the AggLayer or L1, depending on the chosen settlement layer.

Reviewer 4: ```suggestion
- The aggregator sends a "prove-batches" request together with the witness to the prover, and receives a proof in return. The aggregator proof to L1.
```

- Users interact with the bridge service via the bridge UI or API.

> - The AggLayer verifies proofs and interacts with the L1 smart contracts.

Reviewer 4: ```suggestion
- For AggLayer-connected chains, proofs from various chains get aggregated and a single final proof is sent to L1.
```

### User data flow

> The following diagram is a sequential depiction of the user data flow for the CDK FEP stack in validium mode using a mock prover and having an AggLayer connection.

Reviewer 4: It seems this diagram is a bit redundant given the points under the "Component interactions" subsection.

![High level view of CDK user data flow](../../img/cdk/cdk-user-data-flow.svg)

#### Sequential interactions

1. User sends a transaction to the CDK Erigon RPC node.
2. The CDK Erigon RPC node proxies the data to the CDK Erigon sequencer node and syncs the batch data between the sequencer and the RPC nodes.
3. The sequencer sequences the transaction batches.
4. The sequencer sender reads batches from the RPC node.
5. In validium mode only, the sequencer sender persists transaction data into the DAC nodes.
6. The sequencer sender sequences the batches into the L1 smart contracts.
7. The aggregator reads batches from the sequencer data stream.
8. The aggregator sends batches to the provers.
9. The aggregator submits the final proof to the AggLayer.

> 10. The AggLayer submits the final proof to the L1 smart contract domain.

Reviewer 4: It is not clear how this flow differs from the points in the "Component interactions" subsection.

## CDK full execution proof (FEP)

The following diagram depicts the CDK FEP stack component layout and details some of the high level communication between components.

![High level view of CDK stack](../../img/cdk/cdk-stack.png)

### Component interactions

> - Engineers use a CLI tool to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.

Reviewer 5: ```suggestion
- Engineers use a CLI to manage the backend components, installing and initializing various modes of operation, such as validium versus rollup for example.
```

- External applications send transactions to the CDK Erigon RPC node which forwards the transaction data to the sequencer via the transaction pool manager.

> - The transaction pool manager proxies manages the transactions coming in from the RPC node and streams the data onto the sequence sender.

Reviewer 5: Nope, the pool manager do not communicate with the sequence-sender at all, pool manager forwards txs to the sequencer and it's the sequence-sender that reads batches from the data-stream.

Reviewer 5: Next paragraphs are correct

Author: DONE. Pic adjusted also.

- The sequencer sequences transactions batches and puts them into the data streamer.
- The data streamer streams data to the sequencer sender and the aggregator.
- The sequencer sender sequences batches into the L1 smart contract domain and persists data into DAC nodes for validium mode operations.
- The aggregator sends batches to the prover and receives proofs in return. It then aggregates the proofs into batches before submitting them to the AggLayer or L1, depending on the chosen settlement layer.

> - Users interact with the bridge service via the bridge UI.

Reviewer 5: or API

> - The AggLayer also utilizes the bridge service and interacts with the L1 smart contracts.

Reviewer 5: AggLayer doesn't interact with the AggLayer, it only receives proofs from the cdk-node and forwards them to L1

Author: I'm not sure about this @vcastellm ... https://github.com/0xPolygon/polygon-docs/blob/a30e659e665ece8a7be28673842eca335d877363/docs/agglayer/unified-bridge-api.md

Reviewer 5: WDYM? the bridge is not related to the agglayer directly

### User data flow

The following diagram is a sequential depiction of the user data flow for the CDK FEP stack in validium mode using a mock prover and having an AggLayer connection.

![High level view of CDK user data flow](../../img/cdk/cdk-user-data-flow.png)

#### Sequential interactions

> 1. User sends a transaction to the CDK Erigon RPC node.

Reviewer 5: There's a gotcha here: The user can also interact directly with the pool-manager, it's not mandatory to setup an RPC node

2. The Erigon RPC node proxies the data to the CDK Erigon sequencer node.
3. The sequencer sequences the transaction batches.
4. The sequencer puts the batches into the data streamer component.
5. The data streamer component streams the data into the sequence sender. It streams data into the aggregator and AggLayer for rollup and validium mode stacks. 
6. In validium mode, the sequencer sender persists transaction data into the DAC nodes.
7. In both modes, the sequencer sender sequences the batches into the L1 smart contracts.
8. In both modes, the aggregator sends the batches to the prover and the prover returns the proofs.

> 9. The aggregator batches the proofs.

Reviewer 5: ```suggestion
9. The aggregator aggregates the proofs.
```
